### PR TITLE
Laboratory work #3, Ruslana Zobnina - 22FPL2

### PR DESCRIPTION
**Pull request by RuslanaZobnina**

### CODE REVIEW

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]
Good start

File: lab_3_generate_by_ngrams/main.py [hunk 1]
```diff
@@ -41,6 +44,22 @@ def _tokenize(self, text: str) -> Optional[tuple[str, ...]]:
...
+        if not isinstance(text, str):
+            return None
+        tokens = []
+        for token in text.lower():
+            if token.isalpha():
+                tokens.append(token)
+            elif token == ' ' and len(tokens) > 0 and tokens[-1] != self._end_of_word_token:
+                 tokens.append(self._end_of_word_token)
+        if tokens == []:
+                return None
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:56]: too many indents

File: lab_3_generate_by_ngrams/main.py [hunk 2]
```diff
@@ -41,6 +44,22 @@ def _tokenize(self, text: str) -> Optional[tuple[str, ...]]:
...
         """
+        if not isinstance(text, str):
+            return None
+        tokens = []
+        for token in text.lower():
+            if token.isalpha():
+                tokens.append(token)
+            elif token == ' ' and len(tokens) > 0 and tokens[-1] != self._end_of_word_token:
+                 tokens.append(self._end_of_word_token)
+        if tokens == []:
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:55]: bad check - what about truthy/falsy?

File: lab_3_generate_by_ngrams/main.py [hunk 3]
```diff
@@ -41,6 +44,22 @@ def _tokenize(self, text: str) -> Optional[tuple[str, ...]]:
...
+                return None
+
+        if text[-1] != "/?;:'-=+.,£#@$%^><*&()":
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:58]: string.panctuation

File: lab_3_generate_by_ngrams/main.py [hunk 4]
```diff
@@ -76,6 +101,13 @@ def get_token(self, element_id: int) -> Optional[str]:
 
         In case of corrupt input arguments or arguments not included in storage, None is returned
         """
+        if not isinstance(element_id, int) or element_id not in self._storage.values():
+            return None
+
+        for key, value in self._storage.items():
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:107]: filter

File: lab_3_generate_by_ngrams/main.py [hunk 5]
```diff
@@ -93,6 +125,22 @@ def encode(self, text: str) -> Optional[tuple[int, ...]]:
...
+            return None
+
+        tokens = self._tokenize(text)
+        if not tokens:
+            return None
+
+        encoded_text = []
+        for i in tokens:
+            self._put(i)
+            if self.get_id(i) is None:
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:138]: why do you call `get_id` twice ?

File: lab_3_generate_by_ngrams/main.py [hunk 6]
```diff
@@ -159,7 +234,21 @@ def _postprocess_decoded_text(self, decoded_corpus: tuple[str, ...]) -> Optional
 
         In case of corrupt input arguments, None is returned
         """
+        if not isinstance(decoded_corpus, tuple) or not decoded_corpus:
+            return None
+
+        postprocess_decoded_text = ''
+        for index, token in enumerate(decoded_corpus):
+            if index == len(decoded_corpus) - 1:
+                postprocess_decoded_text += '.'
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:243]: fstring

File: lab_3_generate_by_ngrams/main.py [hunk 7]
```diff
@@ -179,7 +268,9 @@ def __init__(self, encoded_corpus: tuple | None, n_gram_size: int) -> None:
             encoded_corpus (tuple): Encoded text
             n_gram_size (int): A size of n-grams to use for language modelling
         """
-
+        self._encoded_corpus = encoded_corpus
+        self._n_gram_size = range(2, 5)
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:272]: what does this range mean?

File: lab_3_generate_by_ngrams/main.py (hunk 3 above)
Comment (RuslanaZobnina) [reply lab_3_generate_by_ngrams/main.py]: done

File: lab_3_generate_by_ngrams/main.py (hunk 4 above)
Comment (RuslanaZobnina) [reply lab_3_generate_by_ngrams/main.py]: done

File: lab_3_generate_by_ngrams/main.py (hunk 5 above)
Comment (RuslanaZobnina) [reply lab_3_generate_by_ngrams/main.py]: fixed

File: lab_3_generate_by_ngrams/main.py (hunk 6 above)
Comment (RuslanaZobnina) [reply lab_3_generate_by_ngrams/main.py]: done

File: lab_3_generate_by_ngrams/main.py (hunk 7 above)
Comment (RuslanaZobnina) [reply lab_3_generate_by_ngrams/main.py]: fixed

File: lab_3_generate_by_ngrams/main.py (hunk 1 above)
Comment (RuslanaZobnina) [reply lab_3_generate_by_ngrams/main.py]: fixed

File: lab_3_generate_by_ngrams/main.py [hunk 8]
```diff
@@ -41,6 +45,21 @@ def _tokenize(self, text: str) -> Optional[tuple[str, ...]]:
         In case of corrupt input arguments, None is returned.
         In case any of methods used return None, None is returned.
         """
+        if not isinstance(text, str):
+            return None
+        tokens = []
+        for token in text.lower():
+            if token.isalpha():
+                tokens.append(token)
+            elif token == ' ' and len(tokens) > 0 and tokens[-1] != self._end_of_word_token:
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:54]: just `not token`

File: lab_3_generate_by_ngrams/main.py [hunk 9]
```diff
@@ -143,6 +203,17 @@ def _decode(self, corpus: tuple[int, ...]) -> Optional[tuple[str, ...]]:
         In case of corrupt input arguments, None is returned.
         In case any of methods used return None, None is returned.
         """
+        if not isinstance(corpus, tuple) or not corpus:
+            return None
+
+        decoded_corpus = []
+        for i in corpus:
+            token = self.get_token(i)
+            if token is None:
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:212]: Why is it specific check for none? not just `not token`?

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

**Pull request closed without merging**